File: ms-inventario/src/main/java/com/smartlogix/ms_inventario/service/ProductoService.java
package com.smartlogix.ms_inventario.service;

import com.smartlogix.ms_inventario.dto.ProductoDTO;
import com.smartlogix.ms_inventario.exception.ResourceNotFoundException;
import com.smartlogix.ms_inventario.model.Producto;
import com.smartlogix.ms_inventario.repository.ProductoRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ProductoService {

    @Autowired
    private ProductoRepository repo;

    public List<Producto> listarTodos() {
        return repo.findAll();
    }

    public List<Producto> listarBajoStock() {
        // Obtenemos todos los productos y filtramos por stock <= stockMinimo
        // O podríamos usar el método del repo si tuviéramos un valor fijo, 
        // pero el stockMinimo es por producto.
        return repo.findAll().stream()
                .filter(p -> p.getStock() <= p.getStockMinimo())
                .toList();
    }

    public Producto obtenerPorId(Long id) {
        return repo.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Producto no encontrado con id: " + id));
    }

    public Producto crearProducto(ProductoDTO dto) {
        Producto p = new Producto();
        p.setNombre(dto.getNombre());
        p.setDescripcion(dto.getDescripcion());
        p.setPrecio(dto.getPrecio());
        p.setStock(dto.getStock());
        p.setStockMinimo(dto.getStockMinimo());
        p.setActivo(true); // Valor por defecto
        return repo.save(p);
    }

    public Producto actualizarProducto(Long id, ProductoDTO dto) {
        Producto p = obtenerPorId(id);
        p.setNombre(dto.getNombre());
        p.setDescripcion(dto.getDescripcion());
        p.setPrecio(dto.getPrecio());
        p.setStock(dto.getStock());
        p.setStockMinimo(dto.getStockMinimo());
        return repo.save(p);
    }

    public void eliminarProducto(Long id) {
        Producto p = obtenerPorId(id);
        p.setActivo(false); // Soft delete
        repo.save(p);
    }

    public Producto ajustarStock(Long id, int cantidad) {
        Producto p = obtenerPorId(id);
        int nuevoStock = p.getStock() + cantidad;
        if (nuevoStock < 0) {
            throw new IllegalArgumentException("El stock resultante no puede ser negativo");
        }
        p.setStock(nuevoStock);
        return repo.save(p);
    }
}

File: ms-inventario/src/productos/producto.service.ts
import { Injectable, BadRequestException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Producto } from './producto.entity';
import { ProductoDTO } from './producto.dto';
import { ResourceNotFoundException } from './exceptions';

@Injectable()
export class ProductoService {
  constructor(
    @InjectRepository(Producto)
    private readonly repo: Repository<Producto>,
  ) {}

  async listarTodos(): Promise<Producto[]> {
    return this.repo.find();
  }

  async listarBajoStock(): Promise<Producto[]> {
    const todos = await this.repo.find();
    return todos.filter((p) => p.stock <= p.stockMinimo);
  }

  async obtenerPorId(id: number): Promise<Producto> {
    const producto = await this.repo.findOne({ where: { id } });
    if (!producto) {
      throw new ResourceNotFoundException(`Producto no encontrado con id: ${id}`);
    }
    // TypeORM bigint can return as string or number depending on driver, but sqlite uses numbers.
    // Let's ensure id is number
    producto.id = Number(producto.id);
    return producto;
  }

  async crearProducto(dto: ProductoDTO): Promise<Producto> {
    const p = new Producto();
    p.nombre = dto.nombre;
    p.descripcion = dto.descripcion || '';
    p.precio = dto.precio;
    p.stock = dto.stock;
    p.stockMinimo = dto.stockMinimo;
    p.activo = true;
    const saved = await this.repo.save(p);
    saved.id = Number(saved.id);
    return saved;
  }

  async actualizarProducto(id: number, dto: ProductoDTO): Promise<Producto> {
    const p = await this.obtenerPorId(id);
    p.nombre = dto.nombre;
    p.descripcion = dto.descripcion || '';
    p.precio = dto.precio;
    p.stock = dto.stock;
    p.stockMinimo = dto.stockMinimo;
    const saved = await this.repo.save(p);
    saved.id = Number(saved.id);
    return saved;
  }

  async eliminarProducto(id: number): Promise<void> {
    const p = await this.obtenerPorId(id);
    p.activo = false;
    await this.repo.save(p);
  }

  async ajustarStock(id: number, cantidad: number): Promise<Producto> {
    const p = await this.obtenerPorId(id);
    const nuevoStock = p.stock + cantidad;
    if (nuevoStock < 0) {
      // In Java, it threw IllegalArgumentException which got mapped to 400 Bad Request.
      // We throw a standard BadRequestException with message.
      throw new BadRequestException('El stock resultante no puede ser negativo');
    }
    p.stock = nuevoStock;
    const saved = await this.repo.save(p);
    saved.id = Number(saved.id);
    return saved;
  }
}
